_transcripts/032-rust-1.48-1.49.md

```diff
@@ -0,0 +1,731 @@
+`cargo test`. At least, I hope you have. And if you have, you might have noticed
+that by default, it doesn't really print anything. Like, if you throw a bunch of
+`println`s in your library or in your tests, nothing really gets printed unless
+the test fails. And that should feel like magic to you, right? Like, if I call
+`println` and my test ran, why doesn't it print? How does this work? And you
+might think like, oh, maybe the testing suite is just like, special, but— and to
+some extent it is. But there's actually a pretty fascinating, I want to call it
+a hack, because it sort of is a hack in the standard library to enable this
+feature, which is the standard library, has a thread-local that the test— it's
+public but unstable and hidden. And it's a hook that the test crate— that is
+what Rust uses to run your tests— it basically sets that thread-local to `true`
+before it runs the test, and then sets it to `false` after. And the print
+function in the standard library. The thing that actually prints to stdout or
+stderr, has logic in it to check this thread-local. And if it's true, it writes
+it to, like, a buffer in memory instead of writing it to the actual stdout.
+
+And this, the fact that this is sort of a little bit of a hack has been visible,
+if you just knew where to dig. One example of this is if you manually open,
+like, stdout or stderr and then write to it, then that will get printed if you
+run `cargo test`. The other is that if you spawn a thread and then print from
+inside the thread, that would still show up in your test output. And now that
+you understand the underlying mechanism, you might see why. It's because there's
+a thread-local that determines whether the standard library is going to be using
+this in-memory buffering or actually write to the real stdout. When you spawn a
+thread, that new thread has no value for that thread-local, and so it just
+defaults to writing to stdout as it would in any sort of normal run of your
+program.
+
+But in 1.49 there's a change that landed, that basically propagates this
+variable out through `thread::spawn`. So if you do a `thread::spawn`, it will
+first check whether that thread-local has been set, and if it has been set, then
+it also sets it in the new thread that you spawned. So this way you now actually
+do get to capture the output, even from spawned threads. There are still ways
+around it, of course, like you can still manually open stdout if you really,
+really need to. Or you can turn off this feature with `cargo test -- --no-
+capture`. But it is interesting to dig into why— what is the underlying
+mechanism that enables the testing suite to do this, and the realization of why
+it didn't work with threads, and why it does now. I just thought that was kind
+of cool.
+
+__Ben__: There are a few library changes. Not very many. There's also a few-
+there's there's two more const-ification ones, to `Poll::is_ready` and
```

Comment:
const-ification?  constification? `const`ification?

Reply:
Judging from https://en.wiktionary.org/wiki/-ification, no `-`. I like the last version, but am happy with any.

Reply:
Ok, `const`ification it is.